src/config-file-watcher.ts

```diff
@@ -0,0 +1,78 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+
+import * as config from './config';
+
+export function activate(context: vscode.ExtensionContext, args: string[]) {
+  let compileCommandsFolder: string|undefined = undefined;
+
+  for (let arg of args) {
+    if (arg.startsWith('--compile-commands-dir')) {
+      let match = arg.match(/--compile-commands-dir="?([^"]*)"?/);
+      compileCommandsFolder = match[1] || undefined;
+    }
+  }
+
+  const watcher = new ConfigFileWatcher(context, compileCommandsFolder);
+}
+
+class ConfigFileWatcher {
+  private databaseWatcher = vscode.workspace.createFileSystemWatcher(
+      '**/{compile_commands.json,compile_flags.txt,.clang-tidy}');
+
+  constructor(private context: vscode.ExtensionContext,
+              private compileCommandsFolder: string|undefined) {
+    this.databaseWatcher.onDidChange(this.handleConfigFilesChanged.bind(this));
+    this.databaseWatcher.onDidCreate(this.handleConfigFilesChanged.bind(this));
+    context.subscriptions.push(this.databaseWatcher);
+  }
+
+  async handleConfigFilesChanged(uri: vscode.Uri) {
+    if ((await vscode.workspace.fs.readFile(uri)).length) {
+      let relativePath = vscode.workspace.asRelativePath(uri, false);
+      let baseName = path.basename(uri.fsPath);
+      let restart = false;
+
+      if (baseName == 'compile_commands.json') {
+        if (this.compileCommandsFolder &&
+            relativePath == path.join(this.compileCommandsFolder, baseName))
+          restart = true;
+        else if (!this.compileCommandsFolder && relativePath == baseName)
+          restart = true;
+      } else {
+        if (relativePath == baseName)
```

Comment:
- Agreed, it doesn't make sense to install a recursive file watcher if you're only interested in the root folder.
- Anyway the compile commands file is often in a `build` subfolder which would be found by clangd but ignored by this code it seems.
- There could also be multiple subfolders for different configs like buildDebug, buildRelease, ... I'm using this together with the CMake Tools copy commands.json to base folder feature to make clangd work without explicit `--compile-commands-dir`.
- Maybe it would be enough to watch only files up to 1 directory level deep in general. Anything else seems out of scope.

Reply:
@hokein @Trass3r the main reason for using a recursive file watcher was because some users asked to support the `compile-commands-dir` option, which meant that the file could be in any subdirectory (even though it could still be made with the RelativePattern method suggested by @hokein).

The other reason however is that when multi folder workspaces are opened, looking only in the workspace.rootPath would not be correct. Using a generic watcher instead notifies a file change in all folders, but the small price to pay is to check if the path is the right one.

> There could also be multiple subfolders for different configs like buildDebug, buildRelease

These however wouldn't be automatically recognised by clangd or would they? (Can we clarify what folders clangd looks for the compilation database except for the root of the project and the `build` folder?)

Of course were we to drop support for the `compile-commands-dir` option the code would be simpler, but I still think that a recursive watcher would be needed.

Reply:
As I said I let CMake Tools copy the commands.json to the base folder after configuration finished.

Reply:
Then I think we don't need to watch for other folders, because the event is triggered when the file is copied in the base folder. Unless clangd looks for compile_commands.json in more folders than I am aware of,

* project's root folder
* `build`

should be the only directories we need to watch into

Reply:
this sounds good to me, it should cover most major cases.

A minor issue is that `$workspacefolder/compile_commands.json` is usually a symlink of `$workspacefolder/build/compile_commands.json`, which means if we change one of them, we will get two prompt-up dialogs, which seems a bit annoying, but I don't have better suggestion.

Reply:
When the events triggers on build/compile_commands.json we could check if it exists a compile_commands.json in the root folder too and if it's the case, simply ignore the event...

Reply:
I wouldn't bother. I'm affected by this too but I'd just pick "yes always" and be done with it.